src/pyhf/pdf.py

```diff
@@ -531,16 +538,20 @@ def logpdf(self, pars, data):
         try:
             tensorlib, _ = get_backend()
             pars, data = tensorlib.astensor(pars), tensorlib.astensor(data)
-            cut = tensorlib.shape(data)[0] - len(self.config.auxdata)
-            actual_data, aux_data = data[:cut], data[cut:]
 
-            mainpdf = self.mainlogpdf(actual_data, pars)
-            constraint = self.constraint_logpdf(aux_data, pars)
+            actual_data = self.main_model._dataprojection(data)
+            aux_data = self.constraint_model._dataprojection(data)
+
+            mainpdf = self.main_model.logpdf(actual_data, pars)
+            constraint = self.constraint_model.logpdf(aux_data, pars)
+
+            result = prob.joint_logpdf([mainpdf, constraint])
 
-            result = tensorlib.sum(tensorlib.stack([mainpdf, constraint]), axis=0)
-            if not self.batch_size:
+            if (
```

Comment:
```suggestion:-0+5
            return result if self.batch_size else tensorlib.reshape(result, (1,))
```

Reply:
this will get dropped anyways in the subsequent PRs